docs/Style.thy

```diff
@@ -0,0 +1,263 @@
+\<close>
+
+section \<open>General Principles\<close>
+
+text \<open>
+Goals:
+  * Readability and consistency:
+      reduce cognitive overhead.
+
+  * Make good use of horizontal space:
+      most screens are wide, not high
+
+  * Semantic indentation:
+      where possible group things according to meaning. But remain syntactically consistent:
+      a sub-term should not go more left than its parent term.
+
+  * Support tools like diff and grep:
+      diff tries to find the "function" each chunk belongs to by indentation. If we remain
+      compatible, diffs are much easier to read. Pull requests mean we read a lot of diffs.
+      grep and other regexp search is not needed as much as it used to, but it's still often
+      useful to be able to find constants and lemmas by regexp search outside of Isabelle
+
+  * Don't fight existing auto-indent where possible:
+      there can be exceptions, but each time we go against defaults, it will be annoying to follow.
+
+  * Compatibility with upstream style where it exists:
+      there is not much consistent Isabelle distribution style, but where there is, we should follow
+      it, so that others can read our code and we can contribute upstream
+\<close>
+
+text \<open>
+The main rules:
+  * Indent your proofs.
+
+  * Follow naming conventions.
+
+  * Be aware that others have to read what you write.
+\<close>
+
+text \<open>
+To not drive proof maintainers insane:
+  * Do not use `auto` except at the end of a proof, see [1].
+
+  * Never check in proofs containing `back`.
+
+  * Instantiate bound variables in preference to leaving schematics in a subgoal
+    (i.e. use erule_tac ... allE in preference to erule allE)
+
+  * Explicitly name variables that the proof text refers to (e.g. with rename_tac)
+
+  * Don't mix object and meta logic in a lemma statement.
+\<close>
+
+section \<open>Indentation\<close>
+
+text \<open>
+Isabelle code is much easier to maintain when indented consistently. In apply style proofs we
+indent by 2 spaces, and add an additional space for every additional subgoal.
+
+In the following example, the rules iffI and conjI add a new subgoal, and fast removes a subgoal.
+The idea is that, when something breaks, the indentation tells you whether a tactic used to solve a
+subgoal or produced new ones.
+\<close>
+
+lemma disj_conj_distribR:
+  "(A \<and> B \<or> C) = ((A \<or> C) \<and> (B \<or> C))"
+  apply (rule iffI)
+   apply (rule conjI)
+    apply fast
+   apply fast
+  apply auto
+  done
+
+text \<open>
+The statement of Hoare triple lemmas are generally indented in the following way.
+\<close>
+
+lemma my_hoare_triple_lemma:
+  "\<lbrace>precondition_one and precondition_two and
+    precondition three\<rbrace>
+   my_function param_a param_b
+   \<lbrace>post_condition\<rbrace>"
+  oops
+
+text \<open> or \<close>
+
+lemma my_hoare_triple_lemma:
+  "\<lbrace>precondition_one and precondition_two
+    and precondition three\<rbrace>
```

Comment:
I think there should be (maybe not in this revision) the concept of wrapping operators in the "operator left" style and "operator right" style. Some people prefer to wrap operations with the operator dangling on the right, e.g.

```
a + b +
c

a + b *
    c
```

while others prefer to wrap the operator and stick it on the left:

```
a + b
+ c

a + b
    * c
```

Where this is particularly useful, and I've been doing it this way for a long while, is in lemma statements (excuse ASCII art):

```
lemma "[| bunch of assumptions;
          more assumptions |]
       ==> conclusion"
```

which allows to clearly see in a large lemma where the conclusion is. Your style guide doesn't disallow that, but it might be worth pointing out that we did not settle this particular argument.
@lsf37 @mbrcknl any further memories/thoughts to add to this, while we're looking at the PR?

Reply:
Perhaps we don't want to try to resolve the argument now. :-)

It would be good to acknowledge that there are two styles, each with advantages and current adherents, and while we currently allow them both, we would prefer people not to constantly reformat each other's code.

That said, in the interests of getting something merged, I'd also be happy to leave this to the next iteration.

If/when we do say something, I would definitely like to make clear that we want the indenting to follow the operator precedence, as in your examples, so:
```
a
+ b
+ c
  * d
  * e
+ f
```
rather than:
```
a
+ b
+ c
* d
* e
```

Reply:
Yes, I'm happy with that (both, leaving it for later, and to explicitly allow both styles)

Reply:
Yeah, that's fine, but I don't want to settle the argument by accident, since I do wrap `==>` and `-->` that way, and I remember from past discussions I'm not alone on that side. So recognition of the stalemate should go in at some point.

Agreed on your point about constantly changing styles, will add to my general principle observation, I think something like that observation should go in early.

Agreed on precedence; it can even be simpler: indent if it's a different operator, but do not mislead audience about precedence, e.g. anything like this is evil:

```
a
* b
  + c
  + d
```

Reply:
Ok, thanks. I'm pretty keen to merge this as starting point and then accumulate this kind of more detailed and more accurate advice through subsequent pull-requests.